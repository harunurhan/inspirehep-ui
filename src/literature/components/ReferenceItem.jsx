import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import { List } from 'antd';
import { Link } from 'react-router-dom';

import AuthorList from './AuthorList';
import Latex from '../../common/components/Latex';
import PublicationInfoList from './PublicationInfoList';

class ReferenceItem extends Component {
  static renderTitle(reference) {
    const recordId = reference.get('control_number');
    const title = reference.getIn(['titles', 0, 'title'], '');
    const label = reference.get('label');

    const labelDisplay = label ? `[${label}] ` : '';

    if (recordId && title) {
      return (
        <Link to={`/literature/${recordId}`}>
          {labelDisplay} <Latex>{title}</Latex>
        </Link>
      );
    }
    return `${labelDisplay}${title}`;
  }

  render() {
    const { reference } = this.props;
    const publicationInfo = reference.get('publication_info');
    return (
      <List.Item>
        <List.Item.Meta
          title={ReferenceItem.renderTitle(reference)}
          description={
            <Fragment>
              <AuthorList
                recordId={reference.get('control_number')}
                authors={reference.get('authors')}
              />
              <PublicationInfoList
                publicationInfo={publicationInfo}
                labeled={false}
              />
            </Fragment>
          }
        />
      </List.Item>
    );
  }
}

ReferenceItem.propTypes = {
  reference: PropTypes.instanceOf(Map).isRequired,
};

export default ReferenceItem;
