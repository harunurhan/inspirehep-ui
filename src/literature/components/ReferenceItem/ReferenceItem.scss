.bulleted-list {
  padding: 0;
  & > li:not(:first-of-type) {
    &:before {
      content: '\2022 ';
    }
  }
}
