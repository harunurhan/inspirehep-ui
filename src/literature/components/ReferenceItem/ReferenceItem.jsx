import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import { List, Row, Col } from 'antd';
import { Link } from 'react-router-dom';

import AuthorsAndCollaborations from '../../../common/components/AuthorsAndCollaborations';
import ArxivEprintList from '../ArxivEprintList';
import Latex from '../../../common/components/Latex';
import PublicationInfoList from '../../../common/components/PublicationInfoList';
import DOIList from '../DOIList';
import URLList from '../URLList';
import './ReferenceItem.scss';
import { LITERATURE } from '../../../common/routes';

class ReferenceItem extends Component {
  static renderLabel(reference) {
    const label = reference.get('label');
    const labelDisplay = label ? `[${label}] ` : null;
    return labelDisplay;
  }

  static renderTitle(reference) {
    const recordId = reference.get('control_number');
    const title = reference.getIn(['titles', 0, 'title']);
    if (recordId && title) {
      return (
        <Link className="f5" to={`${LITERATURE}/${recordId}`}>
          <Latex>{title}</Latex>
        </Link>
      );
    } else if (title) {
      return <div className="f5">{title}</div>;
    }
    return null;
  }

  static renderMisc(reference) {
    const misc = reference.get('misc');
    return misc && <div>{misc}</div>;
  }

  render() {
    const { reference } = this.props;
    const publicationInfo = reference.get('publication_info');
    const arxivEprint = reference.get('arxiv_eprint');
    const dois = reference.get('dois');
    const urls = reference.get('urls');

    const recordId = reference.get('control_number');
    const authors = reference.get('authors');
    const collaborations = reference.get('collaborations');
    const collaborationsWithSuffix = reference.get(
      'collaborations_with_suffix'
    );

    // HACK: `Row` and only single `Col` for the label are used in order to make label and title responsive
    return (
      <List.Item>
        <Row gutter={24} type="flex" justify="start" align="middle">
          <Col>{ReferenceItem.renderLabel(reference)}</Col>
          <List.Item.Meta
            title={ReferenceItem.renderTitle(reference)}
            description={
              <Fragment>
                {ReferenceItem.renderMisc(reference)}
                <AuthorsAndCollaborations
                  recordId={recordId}
                  authors={authors}
                  collaborations={collaborations}
                  collaborationsWithSuffix={collaborationsWithSuffix}
                />
                <ul className="bulleted-inline-list secondary-container">
                  {publicationInfo && (
                    <li className="dib mr1">
                      <PublicationInfoList
                        publicationInfo={publicationInfo}
                        labeled={false}
                        wrapperClassName="di"
                      />
                    </li>
                  )}
                  {arxivEprint && (
                    <li className="dib mr1">
                      <ArxivEprintList
                        eprints={arxivEprint}
                        wrapperClassName="di"
                      />
                    </li>
                  )}
                  {dois && (
                    <li className="dib mr1">
                      <DOIList dois={dois} wrapperClassName="di" />
                    </li>
                  )}
                  {urls && (
                    <li className="dib mr1">
                      <URLList urls={urls} wrapperClassName="di" />
                    </li>
                  )}
                </ul>
              </Fragment>
            }
          />
        </Row>
      </List.Item>
    );
  }
}

ReferenceItem.propTypes = {
  reference: PropTypes.instanceOf(Map).isRequired,
};

export default ReferenceItem;
