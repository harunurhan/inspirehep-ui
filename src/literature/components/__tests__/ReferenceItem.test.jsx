import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import ReferenceItem from '../ReferenceItem';

describe('ReferenceItem', () => {
  it('renders with full reference', () => {
    const reference = fromJS({
      titles: [{ title: 'Title' }],
      arxiv_eprints: [{ value: '123456' }],
      control_number: 12345,
      authors: [{ full_name: 'Author' }],
      publication_info: [
        {
          journal_title: 'Journal',
        },
      ],
      dois: [{ value: '123456.12345' }],
    });
    const wrapper = shallow(<ReferenceItem reference={reference} />);
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders unlinked reference (no control_number)', () => {
    const reference = fromJS({
      titles: [{ title: 'Title' }],
      authors: [{ full_name: 'Author' }],
      arxiv_eprints: [{ value: '123456' }],
      publication_info: [
        {
          journal_title: 'Journal',
        },
      ],
      dois: [{ value: '123456.12345' }],
    });
    const wrapper = shallow(<ReferenceItem reference={reference} />);
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
