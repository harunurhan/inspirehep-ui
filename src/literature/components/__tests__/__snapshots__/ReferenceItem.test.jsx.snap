// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferenceItem renders unlinked reference (no control_number) 1`] = `
<div
  className="ant-list-item"
>
  <Meta
    description={
      <Unknown>
        <AuthorList
          authors={
            Immutable.List [
              Immutable.Map {
                full_name: "Author",
              },
            ]
          }
          enableShowAll={false}
          limit={5}
          recordId={undefined}
          total={-1}
          wrapperClassName={null}
        />
        <PublicationInfoList
          labeled={false}
          publicationInfo={
            Immutable.List [
              Immutable.Map {
                journal_title: "Journal",
              },
            ]
          }
        />
      </Unknown>
    }
    title="Title"
  />
</div>
`;

exports[`ReferenceItem renders with full reference 1`] = `
<div
  className="ant-list-item"
>
  <Meta
    description={
      <Unknown>
        <AuthorList
          authors={
            Immutable.List [
              Immutable.Map {
                full_name: "Author",
              },
            ]
          }
          enableShowAll={false}
          limit={5}
          recordId={12345}
          total={-1}
          wrapperClassName={null}
        />
        <PublicationInfoList
          labeled={false}
          publicationInfo={
            Immutable.List [
              Immutable.Map {
                journal_title: "Journal",
              },
            ]
          }
        />
      </Unknown>
    }
    title={
      <Link
        replace={false}
        to="/literature/12345"
      >
        
         
        <Latex>
          Title
        </Latex>
      </Link>
    }
  />
</div>
`;
